File: incubators/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/adaptor/BasicLocation.java
/*******************************************************************************
 * Copyright (c) 2004 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.adaptor;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import org.eclipse.core.runtime.adaptor.*;
import org.eclipse.osgi.service.datalocation.Location;

/**
 * Internal class.
 */
public class BasicLocation implements Location {
	private static class MockLocker implements Locker {
		public boolean lock() throws IOException {
			// locking always successful
			return true;
		}
		public void release() {
			// nothing to release
		}

	}
	private boolean isReadOnly;
	private URL location = null;
	private Location parent;
	private URL defaultValue;
	private String property;

	// locking related fields
	private File lockFile;
	private Locker locker;
	public static final String PROP_OSGI_LOCKING = "osgi.locking"; //$NON-NLS-1$
	private static String LOCK_FILENAME = ".metadata/.lock"; //$NON-NLS-1$
	public static boolean DEBUG;

	private static boolean isRunningWithNio() {
		try {
			 Class.forName("java.nio.channels.FileLock"); //$NON-NLS-1$
		} catch (ClassNotFoundException e) {
			return false;
		}
		return true;
	}

	public static Locker createLocker(File lock, String lockMode) {
		if (lockMode == null)
			lockMode = System.getProperties().getProperty(PROP_OSGI_LOCKING);
		
		if ("none".equals(lockMode)) //$NON-NLS-1$
			return new MockLocker();
		
		if ("java.io".equals(lockMode)) //$NON-NLS-1$
			return new Locker_JavaIo(lock);
		
		if ("java.nio".equals(lockMode)) { //$NON-NLS-1$
			if (isRunningWithNio()) {
				return new Locker_JavaNio(lock);
			} else {
				// TODO should we return null here.  NIO was requested but we could not do it...
				return new Locker_JavaIo(lock);
			}
		} 
		
		//	Backup case if an invalid value has been specified
		if (isRunningWithNio()) {
			return new Locker_JavaNio(lock);
		} else {
			return new Locker_JavaIo(lock);
		}
	}

	public BasicLocation(String property, URL defaultValue, boolean isReadOnly) {
		super();
		this.property = property;
		this.defaultValue = defaultValue;
		this.isReadOnly = isReadOnly;
	}

	public boolean allowsDefault() {
		return defaultValue != null;
	}

	public URL getDefault() {
		return defaultValue;
	}

	public Location getParentLocation() {
		return parent;
	}

	public synchronized URL getURL() {
		if (location == null && defaultValue != null)
			setURL(defaultValue, false);
		return location;
	}

	public synchronized boolean isSet() {
		return location != null;
	}

	public boolean isReadOnly() {
		return isReadOnly;
	}

	public synchronized boolean setURL(URL value, boolean lock) throws IllegalStateException {
		if (location != null)
			throw new IllegalStateException(EclipseAdaptorMsg.ECLIPSE_CANNOT_CHANGE_LOCATION);
		File file = null;
		if (value.getProtocol().equalsIgnoreCase("file")) //$NON-NLS-1$
			file = new File(value.getFile(), LOCK_FILENAME);
		lock = lock && !isReadOnly;
		if (lock) {
			try {
				if (!lock(file))
					return false;
			} catch (IOException e) {
				return false;
			}
		}
		lockFile = file;
		location = LocationManager.buildURL(value.toExternalForm(), true);
		if (property != null)
			System.getProperties().put(property, location.toExternalForm());
		return lock;
	}

	public synchronized void setParent(Location value) {
		parent = value;
	}

	public synchronized boolean lock() throws IOException {
		if (!isSet())
			return false;
		return lock(lockFile);
	}

	private boolean lock(File lock) throws IOException {
		if (lock == null || isReadOnly)
			return false;

		File parentFile = new File(lock.getParent());
		if (!parentFile.exists())
			if (!parentFile.mkdirs())
				return false;

		setLocker(lock);
		if (locker == null)
			return true;
		boolean locked = false; 
		try {
			locked = locker.lock();
			return locked;
		} finally {
			if (!locked)
				locker = null;
		}
	}

	private void setLocker(File lock) {
		if (locker != null)
			return;
		String lockMode = System.getProperties().getProperty(PROP_OSGI_LOCKING);
		locker = createLocker(lock, lockMode);
	}

	public synchronized void release() {
		if (locker != null)
			locker.release();
	}
}

File: trunk/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/adaptor/BasicLocation.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.adaptor");
Clazz.load (["org.eclipse.core.runtime.internal.adaptor.Locker", "org.eclipse.osgi.service.datalocation.Location"], "org.eclipse.core.runtime.internal.adaptor.BasicLocation", ["java.io.File", "java.lang.IllegalStateException", "org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg", "$.LocationManager", "org.eclipse.core.runtime.internal.adaptor.Locker_JavaIo", "$.Locker_JavaNio"], function () {
c$ = Clazz.decorateAsClass (function () {
this.$isReadOnly = false;
this.location = null;
this.parent = null;
this.defaultValue = null;
this.property = null;
this.lockFile = null;
this.locker = null;
Clazz.instantialize (this, arguments);
}, org.eclipse.core.runtime.internal.adaptor, "BasicLocation", null, org.eclipse.osgi.service.datalocation.Location);
Clazz.pu$h ();
c$ = Clazz.declareType (org.eclipse.core.runtime.internal.adaptor.BasicLocation, "MockLocker", null, org.eclipse.core.runtime.internal.adaptor.Locker);
Clazz.overrideMethod (c$, "lock", 
function () {
return true;
});
Clazz.overrideMethod (c$, "release", 
function () {
});
c$ = Clazz.p0p ();
c$.isRunningWithNio = Clazz.defineMethod (c$, "isRunningWithNio", 
($fz = function () {
try {
Class.forName ("java.nio.channels.FileLock");
} catch (e) {
if (Clazz.instanceOf (e, ClassNotFoundException)) {
return false;
} else {
throw e;
}
}
return true;
}, $fz.isPrivate = true, $fz));
c$.createLocker = Clazz.defineMethod (c$, "createLocker", 
function (lock, lockMode) {
if (lockMode == null) lockMode = System.getProperties ().getProperty ("osgi.locking");
if ("none".equals (lockMode)) return  new org.eclipse.core.runtime.internal.adaptor.BasicLocation.MockLocker ();
if ("java.io".equals (lockMode)) return  new org.eclipse.core.runtime.internal.adaptor.Locker_JavaIo (lock);
if ("java.nio".equals (lockMode)) {
if (org.eclipse.core.runtime.internal.adaptor.BasicLocation.isRunningWithNio ()) {
return  new org.eclipse.core.runtime.internal.adaptor.Locker_JavaNio (lock);
} else {
return  new org.eclipse.core.runtime.internal.adaptor.Locker_JavaIo (lock);
}}if (org.eclipse.core.runtime.internal.adaptor.BasicLocation.isRunningWithNio ()) {
return  new org.eclipse.core.runtime.internal.adaptor.Locker_JavaNio (lock);
} else {
return  new org.eclipse.core.runtime.internal.adaptor.Locker_JavaIo (lock);
}}, "java.io.File,~S");
Clazz.makeConstructor (c$, 
function (property, defaultValue, isReadOnly) {
this.property = property;
this.defaultValue = defaultValue;
this.$isReadOnly = isReadOnly;
}, "~S,java.net.URL,~B");
Clazz.overrideMethod (c$, "allowsDefault", 
function () {
return this.defaultValue != null;
});
Clazz.overrideMethod (c$, "getDefault", 
function () {
return this.defaultValue;
});
Clazz.overrideMethod (c$, "getParentLocation", 
function () {
return this.parent;
});
Clazz.overrideMethod (c$, "getURL", 
function () {
if (this.location == null && this.defaultValue != null) this.setURL (this.defaultValue, false);
return this.location;
});
Clazz.overrideMethod (c$, "isSet", 
function () {
return this.location != null;
});
Clazz.overrideMethod (c$, "isReadOnly", 
function () {
return this.$isReadOnly;
});
Clazz.overrideMethod (c$, "setURL", 
function (value, lock) {
if (this.location != null) throw  new IllegalStateException (org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg.ECLIPSE_CANNOT_CHANGE_LOCATION);
var file = null;
if (value.getProtocol ().equalsIgnoreCase ("file")) file =  new java.io.File (value.getFile (), org.eclipse.core.runtime.internal.adaptor.BasicLocation.LOCK_FILENAME);
lock = lock && !this.$isReadOnly;
if (lock) {
try {
if (!this.lock (file)) return false;
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
return false;
} else {
throw e;
}
}
}this.lockFile = file;
this.location = org.eclipse.core.runtime.adaptor.LocationManager.buildURL (value.toExternalForm (), true);
if (this.property != null) System.getProperties ().put (this.property, this.location.toExternalForm ());
return lock;
}, "java.net.URL,~B");
Clazz.defineMethod (c$, "setParent", 
function (value) {
this.parent = value;
}, "org.eclipse.osgi.service.datalocation.Location");
Clazz.defineMethod (c$, "lock", 
function () {
if (!this.isSet ()) return false;
return this.lock (this.lockFile);
});
Clazz.defineMethod (c$, "lock", 
($fz = function (lock) {
if (lock == null || this.$isReadOnly) return false;
var parentFile =  new java.io.File (lock.getParent ());
if (!parentFile.exists ()) if (!parentFile.mkdirs ()) return false;
this.setLocker (lock);
if (this.locker == null) return true;
var locked = false;
try {
locked = this.locker.lock ();
return locked;
} finally {
if (!locked) this.locker = null;
}
}, $fz.isPrivate = true, $fz), "java.io.File");
Clazz.defineMethod (c$, "setLocker", 
($fz = function (lock) {
if (this.locker != null) return ;
var lockMode = System.getProperties ().getProperty ("osgi.locking");
this.locker = org.eclipse.core.runtime.internal.adaptor.BasicLocation.createLocker (lock, lockMode);
}, $fz.isPrivate = true, $fz), "java.io.File");
Clazz.overrideMethod (c$, "release", 
function () {
if (this.locker != null) this.locker.release ();
});
Clazz.defineStatics (c$,
"PROP_OSGI_LOCKING", "osgi.locking",
"LOCK_FILENAME", ".metadata/.lock",
"DEBUG", false);
});
